### ISSUE TITLE
Current instrumentation strategy yields unacceptable performance

### ISSUE THREAD

**Issue by samestep** [cloverage/cloverage#187]
Consider this [`isqrt`][isqrt] function and contrived `mean-isqrts` function that uses it:
```clojure
(ns example.core)

(defn isqrt [n]
  (let [n (bigint n)]
    (if (zero? n)
      n
      (loop [x (.shiftLeft BigInteger/ONE (quot (inc (.bitLength n)) 2))]
        (let [y (quot (+ x (quot n x)) 2)]
          (if (<= x y)
            x
            (recur y)))))))

(defn mean-isqrts [n]
  (/ (apply + (map isqrt (range n)))
     (double n)))
```
The two functions are tested thusly:
```clojure
(ns example.core-test
  (:require [clojure.test :refer [deftest is]]
            [example.core :refer :all]))

(defmacro measure [f & args]
  `(let [p# (promise)]
     (print (str ~(str f) " " (with-out-str (time (deliver p# (~f ~@args))))))
     @p#))

(deftest isqrt-test
  (is (= 316 (measure isqrt 100000))))

(deftest mean-isqrts-test
  (is (== 210.31854 (measure mean-isqrts 100000))))
```
This is very similar to a subset of an actual namespace for which I am trying to use Cloverage, and it illustrates the problem I'm having.

Currently, Cloverage's instrumentation strategy is to first instrument all functions that need to be instrumented, then run all tests. This results in unacceptable performance:
```
$ lein test

lein test example.core-test
mean-isqrts "Elapsed time: 86.613774 msecs"
isqrt "Elapsed time: 0.023023 msecs"

Ran 2 tests containing 2 assertions.
0 failures, 0 errors.
```
```
$ lein cloverage --nop
WARNING: reader-conditional already refers to: #'clojure.core/reader-conditional in namespace: clojure.tools.reader.impl.utils, being replaced by: #'clojure.tools.reader.impl.utils/reader-conditional
WARNING: tagged-literal already refers to: #'clojure.core/tagged-literal in namespace: clojure.tools.reader.impl.utils, being replaced by: #'clojure.tools.reader.impl.utils/tagged-literal
Loading namespaces:  (example.core)
Test namespaces:  (example.core-test)
Loaded  example.core  .
Instrumented namespaces.

Testing example.core-test
mean-isqrts "Elapsed time: 317.108315 msecs"
isqrt "Elapsed time: 0.026342 msecs"

Ran 2 tests containing 2 assertions.
0 failures, 0 errors.
Ran tests.
```
```
$ lein cloverage
WARNING: reader-conditional already refers to: #'clojure.core/reader-conditional in namespace: clojure.tools.reader.impl.utils, being replaced by: #'clojure.tools.reader.impl.utils/reader-conditional
WARNING: tagged-literal already refers to: #'clojure.core/tagged-literal in namespace: clojure.tools.reader.impl.utils, being replaced by: #'clojure.tools.reader.impl.utils/tagged-literal
Loading namespaces:  (example.core)
Test namespaces:  (example.core-test)
Loaded  example.core  .
Instrumented namespaces.

Testing example.core-test
mean-isqrts "Elapsed time: 5762.560537 msecs"
isqrt "Elapsed time: 0.111505 msecs"

Ran 2 tests containing 2 assertions.
0 failures, 0 errors.
Ran tests.
```
In my actual code, I use [`test.check`][test.check] alongside my example-based tests, so the execution time is even more pronounced; this example serves to show that it is an execution time problem caused specifically by instrumentation, though, as the relatively fast times from using `--nop` show that this isn't just a symptom of #102.

It isn't a solution to move the performance-critical functions into their own namespaces and exclude those namespaces from instrumentation, because I still want to get coverage information on those functions.

In my opinion it would be better to be able to mark tests as testing specific functions:
```clojure
(deftest ^{:testing #'isqrt} isqrt-test
  (is (= 316 (isqrt 100000))))
```
Cloverage could first do the following for each function that is mentioned in the `:testing` metadata of at least one test:
- Instrument the function
- Run the tests that mark themselves as `:testing` that function
- Unstrument the function (to borrow a term from [`clojure.spec`][unstrument])

Then Cloverage would instrument all the functions that weren't mentioned and run all the remaining tests.

[isqrt]: https://codereview.stackexchange.com/q/120409/82369
[test.check]: https://github.com/clojure/test.check
[unstrument]: https://clojure.github.io/clojure/branch-master/clojure.spec-api.html#clojure.spec.test/unstrument

**Comment by rm-hull** [comment]
I've never heard of any other coverage tooling do this kind of thing. 

Generally (for me at least) running coverage is an agglomerative action where I want a summary of the entire coverage across the whole codebase after running the test suite - not just some specific functions under test.

Maybe it would be good to look at the instrumentation code first to see if there are any obvious bottlenecks that can be eliminated in the implementation as-is, rather than attempting a more complex solution.

**Comment by samestep** [comment]
> Generally (for me at least) running coverage is an agglomerative action where I want a summary of the entire coverage across the whole codebase after running the test suite - not just some specific functions under test.

I agree, I use it the same way and I have no intention of changing the way Cloverage is run or the output it produces. I hope that I haven't been communicating poorly about what my proposed solution is and why I think it is viable; let me try to clarify.

My thinking is that for each performance-sensitive function, there should already be tests that provide the necessary coverage of that function, and that the additional coverage contributed by tests of other functions that call those lower-level functions should be negligible. Therefore, I'm proposing that Cloverage still runs all the tests, but rather than instrumenting all the functions for all the tests, allow the developer to tell Cloverage not to instrument some functions when running tests where the additional coverage of those functions wouldn't matter.

Then after running all these tests, some with different functions instrumented than others, Cloverage would still merge together all the coverage results and produce output in the same way. In this way, the coverage it reports would be a lower bound rather than an exact measure, but if the developer has specific tests written for any of these performance-critical functions (which, indeed, is required by the very design I propose above), it should be a relatively close lower bound.

> Maybe it would be good to look at the instrumentation code first to see if there are any obvious bottlenecks that can be eliminated in the implementation as-is, rather than attempting a more complex solution.

In my actual code, running the tests by themselves for the offending namespace takes a few seconds, while with Cloverage, running one small fraction of one of the tests takes around a minute. Since I'm seeing several orders of magnitude of difference in performance, I suspect that my specific problem cannot be solved by this sort of optimization (although I could be wrong).

In general, though, it seems to me that it's always possible to write very low-level bits of code where each form is meant to have a very small execution time, and thus it seems like it should always be possible to write functions where the runtime is dominated by Cloverage instrumentation.

I'll implement a prototype of my proposal so we can have something concrete to discuss.

**Comment by samestep** [comment]
@rm-hull Update: earlier today I was working on implementing a prototype, and I realized that the instrumentation works differently than I expected. I had thought that it would be relatively straightforward to do the sort of strategic instrumentation/unstrumentation that I described in my earlier post, but it looks like that's not the case.

I still think it would be possible and preferable to implement and use a solution like the one I proposed above; however, I thought a bit more about my specific codebase, and have decided to take the less preferable "move the performance-critical functions into their own namespaces and exclude those namespaces from instrumentation" approach I mentioned above, and just accept the downside of not having coverage data for those functions.

Let me know if you think it would be a good idea for me to continue working on the prototype; if not, I'll probably shelve the idea until I have a need for it in the future.
